accord-core/src/main/java/accord/primitives/Txn.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.primitives;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import accord.local.Command;
+import accord.local.CommandStore;
+
+import accord.api.*;
+import accord.local.SafeCommandStore;
+import accord.utils.ReducingFuture;
+import org.apache.cassandra.utils.concurrent.Future;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public interface Txn
+{
+    enum Kind
+    {
+        READ, WRITE;
+
+        public boolean isWrite()
+        {
+            return this == WRITE;
+        }
+    }
+
+    class InMemory implements Txn
+    {
+        private final Kind kind;
+        private final Keys keys;
+        private final Read read;
+        private final Query query;
+        private final Update update;
+
+        public InMemory(@Nonnull Keys keys, @Nonnull Read read, @Nonnull Query query)
+        {
+            this.kind = Kind.READ;
+            this.keys = keys;
+            this.read = read;
+            this.query = query;
+            this.update = null;
+        }
+
+        public InMemory(@Nonnull Keys keys, @Nonnull Read read, @Nonnull Query query, @Nullable Update update)
+        {
+            this.kind = Kind.WRITE;
+            this.keys = keys;
+            this.read = read;
+            this.update = update;
+            this.query = query;
+        }
+
+        protected InMemory(@Nonnull Kind kind, @Nonnull Keys keys, @Nonnull Read read, @Nullable Query query, @Nullable Update update)
+        {
+            this.kind = kind;
+            this.keys = keys;
+            this.read = read;
+            this.update = update;
+            this.query = query;
+        }
+
+        @Override
+        public PartialTxn slice(KeyRanges ranges, boolean includeQuery)
+        {
+            return new PartialTxn.InMemory(
+                    ranges, kind(), keys().slice(ranges),
+                    read().slice(ranges), includeQuery ? query() : null,
+                    update() == null ? null : update().slice(ranges)
+            );
+        }
+
+        @Override
+        public Kind kind()
+        {
+            return kind;
+        }
+
+        @Override
+        public Keys keys()
+        {
+            return keys;
+        }
+
+        @Override
+        public Read read()
+        {
+            return read;
+        }
+
+        @Override
+        public Query query()
+        {
+            return query;
+        }
+
+        @Override
+        public Update update()
+        {
+            return update;
+        }
+
+        @Override
+        public boolean equals(Object o)
```

Comment:
We do not store this object in a hash collection, and if we ever were to it would be nonsensical to do so based on equality rather than object identity. It is misleading therefore to implement the method, as it suggests we do store this in a hash collection.

Reply:
Then implement hashCode() to throw UnsupportedOperationException?

Your point is valid, but I don't think we should leave it at them disagreeing.

Reply:
I could imagine equals might be useful/used for tests?

Sure, I'm happy to throw UOE, it makes sense to me.